File: lib/math.cpp
#include "number.h"

#include <iso646.h>

#include "intrinsic.h"
#include "rtl_exec.h"

namespace rtl {

namespace ne_math {

Number _CONSTANT_PRECISION_DIGITS()
{
    return number_from_uint32(34);
}

Number abs(Number x)
{
    return number_abs(x);
}

Number acos(Number x)
{
    return number_acos(x);
}

Number acosh(Number x)
{
    return number_acosh(x);
}

Number asin(Number x)
{
    return number_asin(x);
}

Number asinh(Number x)
{
    return number_asinh(x);
}

Number atan(Number x)
{
    return number_atan(x);
}

Number atanh(Number x)
{
    return number_atanh(x);
}

Number atan2(Number y, Number x)
{
    return number_atan2(y, x);
}

Number cbrt(Number x)
{
    return number_cbrt(x);
}

Number ceil(Number x)
{
    return number_ceil(x);
}

Number cos(Number x)
{
    return number_cos(x);
}

Number cosh(Number x)
{
    return number_cosh(x);
}

Number erf(Number x)
{
    return number_erf(x);
}

Number erfc(Number x)
{
    return number_erfc(x);
}

Number exp(Number x)
{
    return number_exp(x);
}

Number exp2(Number x)
{
    return number_exp2(x);
}

Number expm1(Number x)
{
    return number_expm1(x);
}

Number floor(Number x)
{
    return number_floor(x);
}

Number frexp(Number x, Number *exp)
{
    int iexp;
    Number r = number_frexp(x, &iexp);
    *exp = number_from_sint32(iexp);
    return r;
}

Number hypot(Number x, Number y)
{
    return number_hypot(x, y);
}

Number intdiv(Number x, Number y)
{
    return number_trunc(number_divide(x, y));
}

Number ldexp(Number x, Number exp)
{
    if (not number_is_integer(exp)) {
        // TODO: more specific exception?
        throw RtlException(ne_global::Exception_ValueRangeException, utf8string(number_to_string(exp)));
    }
    return number_ldexp(x, number_to_sint32(exp));
}

Number lgamma(Number x)
{
    return number_lgamma(x);
}

Number log(Number x)
{
    return number_log(x);
}

Number log10(Number x)
{
    return number_log10(x);
}

Number log1p(Number x)
{
    return number_log1p(x);
}

Number log2(Number x)
{
    return number_log2(x);
}

Number max(Number a, Number b)
{
    return ne_global::max(a, b);
}

Number min(Number a, Number b)
{
    return ne_global::min(a, b);
}

Number nearbyint(Number x)
{
    return number_nearbyint(x);
}

bool odd(Number x)
{
    return ne_global::odd(x);
}

Number powmod(Number b, Number e, Number m)
{
    if (not number_is_integer(b)) {
        throw RtlException(ne_global::Exception_ValueRangeException, utf8string(number_to_string(b)));
    }
    if (not number_is_integer(e)) {
        throw RtlException(ne_global::Exception_ValueRangeException, utf8string(number_to_string(e)));
    }
    if (not number_is_integer(m)) {
        throw RtlException(ne_global::Exception_ValueRangeException, utf8string(number_to_string(m)));
    }
    return number_powmod(b, e, m);
}

Number round(Number places, Number value)
{
    return ne_global::round(places, value);
}

Number sign(Number x)
{
    return number_sign(x);
}

Number sin(Number x)
{
    return number_sin(x);
}

Number sinh(Number x)
{
    return number_sinh(x);
}

Number sqrt(Number x)
{
    return number_sqrt(x);
}

Number tan(Number x)
{
    return number_tan(x);
}

Number tanh(Number x)
{
    return number_tanh(x);
}

Number tgamma(Number x)
{
    return number_tgamma(x);
}

Number trunc(Number x)
{
    return number_trunc(x);
}

} // namespace ne_math

} // namespace rtl

File: rtl/jvm/neon/Math.java
package neon;

import java.math.BigDecimal;

public class Math {

    public static neon.type.Number abs(neon.type.Number x) {
        return new neon.type.Number(x.repr.abs());
    }

    public static neon.type.Number acos(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.acos(x.repr.doubleValue()));
    }

    public static neon.type.Number asin(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.asin(x.repr.doubleValue()));
    }

    public static neon.type.Number atan(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.atan(x.repr.doubleValue()));
    }

    public static neon.type.Number ceil(neon.type.Number x) {
        return new neon.type.Number(x.repr.setScale(0, BigDecimal.ROUND_CEILING));
    }

    public static neon.type.Number cos(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.cos(x.repr.doubleValue()));
    }

    public static neon.type.Number exp(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.exp(x.repr.doubleValue()));
    }

    public static neon.type.Number floor(neon.type.Number x) {
        return new neon.type.Number(x.repr.setScale(0, BigDecimal.ROUND_FLOOR));
    }

    public static neon.type.Number log(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.log(x.repr.doubleValue()));
    }

    public static neon.type.Number intdiv(neon.type.Number x, neon.type.Number y) {
        return new neon.type.Number(x.repr.divideToIntegralValue(y.repr));
    }

    public static neon.type.Number max(neon.type.Number a, neon.type.Number b)
    {
        if (a.compareTo(b) > 0) {
            return a;
        } else {
            return b;
        }
    }

    public static neon.type.Number min(neon.type.Number a, neon.type.Number b)
    {
        if (a.compareTo(b) < 0) {
            return a;
        } else {
            return b;
        }
    }

    public static Boolean odd(neon.type.Number x)
    {
        return !x.repr.remainder(new BigDecimal(2)).equals(BigDecimal.ZERO);
    }

    public static neon.type.Number sign(neon.type.Number x) {
        return new neon.type.Number(x.repr.signum());
    }

    public static neon.type.Number sin(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.sin(x.repr.doubleValue()));
    }

    public static neon.type.Number sqrt(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.sqrt(x.repr.doubleValue()));
    }

    public static neon.type.Number tan(neon.type.Number x) {
        return new neon.type.Number(java.lang.Math.tan(x.repr.doubleValue()));
    }

    public static neon.type.Number trunc(neon.type.Number x) {
        return new neon.type.Number(x.repr.divide(BigDecimal.ONE, BigDecimal.ROUND_DOWN));
    }

}
